source/gfoidl.Base64/Internal/Base64UrlEncoder/Base64UrlEncoder.Encode.cs

```diff
@@ -231,8 +231,30 @@ private static int GetNumBase64PaddingCharsAddedByEncode(int dataLength)
             // 0 -> 0
             // 1 -> 2
             // 2 -> 1
+            uint mod3 = FastMod3((uint)dataLength);
+            return mod3 == 0 ? 0 : (int)(3 - mod3);
```

Comment:
It would be nice if the JIT can emit `cmov` here, so it gets branchless. Unfortunately there is no "trick" to force the JIT to do so, except some very obscure hacks (which we shouldn't use here).

Reply:
I am curious to know the obscure hacks...

Reply:
Not exactely a `cmov` (which would be desired), but still a branchless version with some bit-twiddling and the knowledge that `true / false` is represented as `1 / 0` -> [snippet in sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAMyUATOS7kA3svJHKG1i3IBZABRmM5CdkEMYASnKHjBpcZ+UA7OQAQlDYTGAAFsK4uOLCYBhWDk4w5AB8ALzkpGjkKJo5cFQuit7GAL4eRpXkANoWMBjhEAAmAJL8glb1jS3tfIIA8nxsEEy4NACCAOZTsNGsEjCtTIJmZlMuALrVfFAL2Bgp1EjktkEhYZGyMTBxCcAQEILkkEzNrCNMOWdJzq0AZgAVKDOb7mX5Lf4AMUcuFc1S8vmMxACViscAAqmNsP9OBNcAAeB5PMEYVJWWD/F6jd6fFxuABk5ESjj+QJBKTg9lZkJhgjh9PIAGpuckAXy4SUfBUlGUgA)
It the stack isn't used, it would be even better.

The main advantage will be when the branch-predictor can't work reliable due to "random" inputs, i.e. unknown patters for `mod3 == 0`. Then this approach may gain some perf.
Contrast this with [clang](https://godbolt.org/z/aztkkn), so it could be pretty concise branchless code.

Reply:
Interesting: if the `bool` is "casted" to `byte` instead of `int` the stack-usage goes away and code looks [pretty good](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAMyUATOS7kA3svJHKG1i3IBZABRmM5CdkEMYASnKHjBpcZ+UA7OQAQlDYTGAAFsK4uOLCYBhUVg5OMOQAfAC85KRo5CiauXBULorexgC+HkZV5ADaFjAY4RAAJgCS/IJWDU2tHXyCAPJ8bBBMuDQAggDm07DRrBIwbUyCZmbTLgC6NXxQi9gYqdRI5LZBIWGRsjEwcRhWwBAQguSQTC2so0y558nObQAZgAVKDOX7mf7LQEAMUcuFcNS8vmMxACViscAAquNsIDOJNcAAeJ4vCEYNJWWCAt5jT7fFxuABk5CSjgBILBqTg9nZ0LhggRjPIAGpeSkgQKEaUfJUytV5XUes12p1uo0Vf0hiNWGMJjM5jdFstVusmJsdoq9gcjpQqKdzsFQhEord7olSa93vTdT8zpC+UDQeD/XYoZL4YjFciUf5WZicbg8QTiZ7csAAJ5HSnU2kfL6+4UstkSznOcg88OwyPCsVVqUwGUVZTlIA==).

Can you give this a try with the above benchmark? If it has potential to be faster, we should go with such a approach (although state otherwise above :wink:).

Reply:
Tried with the [`BranchlessSelect`](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzMMA2gFkYGABYQAJgElxkgBQ37T12MkB5MT4ILlwWAEEAc0jYXFxeBRhnLkleLjTIgEoAXVNzHTM0jAYAcVsAUS5IRxhHABkYLkj7AH0AORgAdxaAIShsKrtpOJa2GESoXBh3IoZcCA4oFQam+0y8sxNqcx2GAHo9hlkHKAwwDmKAM2hSjl5HBgGHiHsZBnIkBmAATwwYBkc2Aw2A2u14lwY7nmi2WjWadgYAF5Ee8kOttrtMcQAOwMUikDTUUE7WbAPAwVCVaq1FZI0q2HrkylVJw0uH2KELJYwFbwzKEzGFHgMLgcfAABWwjkcGSEdmwkzpZQwbTFjKmqEl0tl8sm4WltR63yprJaADE8BgrE5ev1BsNcKNxjIppyYTz2XZ+ZoMVjcWSNSgTTV6o0GHARWKtTKmnKFepQQBfcyg6y2BwuNyedM+NyBYKhCLRWLxRLJVLpJo5UEFBizZXBtmrOztLq2gZgIaqXAzYXQ7m8tagraC/aHY7QM4XBjXKC3e6PLjPV5zj5fX7/QHA4nmcGQ/uw5tIlEfdGjnY4vEEn2j0lMoMskO0lHK9UUh/U0PNt0Dz3e31gsKooSlKMaRHGiovrYqr4G+mqgTq8b6iGRqNualrWo47b2t2P6HnyhI7mYl4Bu+jZfuGkYgdqsa6gmAHJmYqZeBmvgeCxuZ+PmvAhGEUQxN2CRJCkaQZNWAG1vWFSPk28Ktp0vbFAeHrNme5gjoKBxHCcU5XDcJR3A8TwMC8dhvGuPx/ACQIggBJIQnhKnwseqJqeeTC4vihF2buwqkcyn7PvSGBwR+rJfvCjmDl6AqCrMwHRohkHBTBoWJbRSEGo4qEyehuBWk4UV/t5o4kfe5G0hGCUIRlkyxWYibMTmmZ+Nm3gtQEQQ8YW/ElkJ5aiVWuQAWIUAvDAYB/A8zCfFJIXlTJEUcrMynRW5xhEXWQFigA6g40j+FAkqnLw2CSMkYgXD0kgQGAADWuB0kVR4ANR4pk+wMNo9W7JewF7RAB1HQqfBnRdV03fdj0AFQMCgP2MUxRIAWm7VsW1rF5l1vFFgJcT9SJlZZMNOyjQkQL/NwuDYJc/wzVtxTKql97peBdHIYaxq5Ra+WYYp1nAmtw6bVpQhneckhAt1da4CAIuHLgnS8GcCLuFu2DRQwACkX3rTsWmMHAAB8DB0PL7zhibpDm/oxvvJtdzCvgTjfZt4LuJUShjVwAg8Cwzi4KgPTK+KY0qHEevqZtOzO44ugojzBVx+47iOxgmTq2tP07I1PlmDAkhTJtGnubH8eQmnGc2ZrOuu3nDCIzegqXmXDAAIQoowAD8kJFJk7i6BGZcfSApsI5iTVo1mHEddxOO9YJZaE2JJPmGTCgUwwVM03Ta5zczgasxBuAc9lXPUnlSfYZ2Dr85nf7C/XovixwksFjLctPwrSsq5C99HrXSOZgDaW1NubcgoDrZfzxKAqg9c04MDLtnXclwPZcC9iEX2GB/aBxQMHDAodbrdiARteuMcXZ0kTphbQKdK7/wIptXOo4C5F3riXc8rcUSpz7vQ+w2svrIIaimZGpVcR9A7F2OI8hpCTXcJwzuaBTaKMUv3QeiCXaZH/DnSemNWozzYnPHqxZF7CQrCvUE69N7b1pkwPewomZqhZjVNmmUULn1NFQm04icIjDGBMV0sxeFekfqOZ+kgJZSxCB/c2itladj/tXT0/DtAkJAXbM20CIF2ygaEw4tsTZwNHAgpBbtUGe14N7LBOCg4hzDsQ4u0dzCcIYJ45O3CeBV0FsVRhm0WEwAaWQppFCuF0MSQAgRjDhGbUvN4m+3ZpETQwHIihHclG9w6QPSiw9FF0C0eYJh5hUa6PYs1Ax2MjF41LKYwaxMLFQHJlZemsxZmSNwAs2RwAICAwYJAJcytuqKNmBvSQHAkiXAACpQFBYC4UwLQXOEuBaQuMB1rsIvLiFOcAACqoQd4RFwAAHk+YDRRlkYBG3cLACEvyZQFk0QwAAZJCOFYLIWgsoiyhFSKpj0repyxFZ0pgIybmYI5nETlTy4ucvixj8ZLzMUNGsuh6YINaTQhBLLUXmx2pTKY6iZSXF4LUBgAARAYRrJAMAaPgCpMAADkj1LiWljo8SQkRoDKzsPgSEdgMAYDELLA4kRPUcGACwSA+A9iOBeFwWwewoU8H4DAPYl1JCSD2CgOgaI0Dm06HYZQCIzo3U6I9DAEBHjEBQAwV4DAbpNEQa/PgfhlCRK4HWCENaZRKHmHOXgj0HTVvlG20g0NobaFYA7OtkRa0QE6MAZWj0RlTpYFYbACAABqZ12WHF0G9cgH1YYspKppQ4qB53FFhme5WSICmkAABznpllwO1xRzhQFgDwSQ3w5gcDEGIScxqfgMFXXw31/rA17GDfYUN4bRBRpjXGyAVKEB7D7bgUFuAM3kDvfic2pkZBKz1Z0e1DxOgDGKPYPt1by3AH+O09OIG7AsGDpEKwr93DFvPbgRRXa+3QEUQsYoLQSEIPzZEOwXGnqVxTinTjC6GBGxNtoUgH090Ht1gppTKnj1+gYGJiTC7hXUETEAA===) casted as byte. The results are interesting:

|                      Method | Value |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |
|---------------------------- |------ |---------:|----------:|----------:|---------:|------:|--------:|
|                        **Mod3** |    **64** | **2.607 ns** | **0.0991 ns** | **0.2890 ns** | **2.519 ns** |  **1.00** |    **0.00** |
|                    FastMod3 |    64 | 2.208 ns | 0.0491 ns | 0.1351 ns | 2.165 ns |  0.85 |    0.11 |
|         FastMod3_Branchless |    64 | 2.611 ns | 0.0502 ns | 0.0445 ns | 2.606 ns |  1.07 |    0.06 |
| FastMod3_BranchlessReversed |    64 | 2.299 ns | 0.0066 ns | 0.0051 ns | 2.300 ns |  0.94 |    0.05 |
|                             |       |          |           |           |          |       |         |
|                        **Mod3** |    **65** | **2.357 ns** | **0.0494 ns** | **0.0462 ns** | **2.350 ns** |  **1.00** |    **0.00** |
|                    FastMod3 |    65 | 2.108 ns | 0.0411 ns | 0.0365 ns | 2.094 ns |  0.89 |    0.02 |
|         FastMod3_Branchless |    65 | 2.622 ns | 0.0626 ns | 0.0615 ns | 2.607 ns |  1.11 |    0.04 |
| FastMod3_BranchlessReversed |    65 | 2.313 ns | 0.0467 ns | 0.0437 ns | 2.298 ns |  0.98 |    0.02 |
|                             |       |          |           |           |          |       |         |
|                        **Mod3** |    **66** | **2.157 ns** | **0.0346 ns** | **0.0289 ns** | **2.166 ns** |  **1.00** |    **0.00** |
|                    FastMod3 |    66 | 1.920 ns | 0.0396 ns | 0.0471 ns | 1.918 ns |  0.89 |    0.02 |
|         FastMod3_Branchless |    66 | 2.637 ns | 0.0588 ns | 0.0577 ns | 2.631 ns |  1.23 |    0.04 |
| FastMod3_BranchlessReversed |    66 | 2.334 ns | 0.0617 ns | 0.0606 ns | 2.327 ns |  1.09 |    0.03 |

The 'branchless' is not as efficient as expected .
When reverting the branch `BranchlessSelect(mod3 != 0, 3 - mod3, 0)` instead of `BranchlessSelect(mod3 == 0, 0, 3 - mod3)`, the result are better but still under the branched one.

Reply:
Thanks for the numbers. 👍 

Some thoughts -- but I don't know what matches here...

The effect of `cmov` et.al. depends on the distribution of taken / not taken branches, i.e. on how effecient the branch predictor can work. If the branch predictor does a good job, then the "branchy code" may be more efficient, as the pipeline can speculatively process the following steps, whereas with `cmov` the rest of the pipeline needs to wait for the result of `cmov`.

The code for the branchless version is quite a bit, so it may be more efficient just to use simple and compact code with a branch -- which can be predicted, as in the benchmarks where for constant input length the result will always be the same.
I'm pretty sure that there will be a benchmark constellation where the branchless version will be quite a bit faster.

In effect I think we should go with the code as is. It's easy and not hacky...and as the numbers show quite good. Furthermore I doubt a slight improvement here will show any effect on overall perf.

Once the runtime (JIT) will detect a pattern like `return cond ? a : b` and emit better code (maybe with tiering) we'll gain the perf-win for free.

Reply:
More or less just for fun I tried a vectorized approach to make it branchless. I show it here because you are curios :wink:

Benchmark is laid out so that (hopefully) the branch-predictor won't do a good job.
```
|      Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|------------ |----------:|----------:|----------:|------:|--------:|
|     Default | 1.0953 ns | 0.0213 ns | 0.0199 ns |  1.00 |    0.00 |
|  Branchless | 0.8785 ns | 0.0136 ns | 0.0127 ns |  0.80 |    0.02 |
|  Vectorized | 0.9657 ns | 0.0210 ns | 0.0186 ns |  0.88 |    0.02 |
| Vectorized1 | 0.8840 ns | 0.0156 ns | 0.0146 ns |  0.81 |    0.02 |
```

<details>
  <summary>Benchmark-code</summary>

```c#
using System;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;
using BenchmarkDotNet.Attributes;

#if !DEBUG
using BenchmarkDotNet.Running;
#endif

namespace ConsoleApp4
{
    class Program
    {
        static void Main(string[] args)
        {
            var bench = new Bench();

            int mod3 = 0;
            Console.WriteLine(bench.Default(mod3));
            Console.WriteLine(bench.Branchless(mod3));
            Console.WriteLine(bench.Vectorized(mod3));
            Console.WriteLine(bench.Vectorized1(mod3));
            Console.WriteLine();

            mod3 = 1;
            Console.WriteLine(bench.Default(mod3));
            Console.WriteLine(bench.Branchless(mod3));
            Console.WriteLine(bench.Vectorized(mod3));
            Console.WriteLine(bench.Vectorized1(mod3));

            Console.WriteLine();

            mod3 = 2;
            Console.WriteLine(bench.Default(mod3));
            Console.WriteLine(bench.Branchless(mod3));
            Console.WriteLine(bench.Vectorized(mod3));
            Console.WriteLine(bench.Vectorized1(mod3));

#if !DEBUG
            BenchmarkRunner.Run<Bench>();
#endif
        }
    }

    public class Bench
    {
        [Benchmark(Baseline = true, OperationsPerInvoke = 3)]
        public int Default()
        {
            int s = 0;

            for (int i = 0; i < 3; ++i)
            {
                s += this.Default(i);
            }

            return s;
        }

        [Benchmark(OperationsPerInvoke = 3)]
        public int Branchless()
        {
            int s = 0;

            for (int i = 0; i < 3; ++i)
            {
                s += this.Branchless(i);
            }

            return s;
        }

        [Benchmark(OperationsPerInvoke = 3)]
        public int Vectorized()
        {
            int s = 0;

            for (int i = 0; i < 3; ++i)
            {
                s += this.Vectorized(i);
            }

            return s;
        }

        [Benchmark(OperationsPerInvoke = 3)]
        public int Vectorized1()
        {
            int s = 0;

            for (int i = 0; i < 3; ++i)
            {
                s += this.Vectorized1(i);
            }

            return s;
        }

        public int Default(int mod3)    => mod3 == 0 ? 0 : 3 - mod3;
        public int Branchless(int mod3) => BranchlessSelect(mod3 != 0, 3 - mod3, 0);

        public int Vectorized(int mod3)
        {
            Vector128<int> vMod3       = Vector128.CreateScalarUnsafe(mod3);
            Vector128<int> zero        = Vector128<int>.Zero;
            Vector128<int> valueIfTrue = Vector128.CreateScalarUnsafe(3 - mod3);

            Vector128<int> eqZero  = Sse2.CompareEqual(vMod3, zero);
            Vector128<int> neqZero = Sse2.CompareEqual(eqZero, zero);
            Vector128<int> result  = Sse2.And(neqZero, valueIfTrue);

            return Sse2.ConvertToInt32(result);
        }

        public int Vectorized1(int mod3)
        {
            Vector128<int> vMod3       = Vector128.CreateScalarUnsafe(mod3);
            Vector128<int> zero        = Vector128<int>.Zero;
            Vector128<int> valueIfTrue = Vector128.CreateScalarUnsafe(3 - mod3);

            Vector128<int> eqZero = Sse2.CompareEqual(vMod3, zero);
            Vector128<int> result = Sse2.AndNot(eqZero, valueIfTrue);

            return Sse2.ConvertToInt32(result);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int BranchlessSelect(bool condition, int valueIfTrue, int valueIfFalse)
        {
            return ((-Unsafe.As<bool, byte>(ref condition)) & (valueIfTrue - valueIfFalse)) + valueIfFalse;
        }
    }
}

```
</details>

[Codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ovLrmXrNjZu258BLIxhNmLLABoAOJBuqaAzEykDEIMAN40DFHRUcSBphgMACIwAGbYHJIYABQJDPgQACb+AJSRMdER1BU1TADs+UWBALzNDIwA/O0MIAyBcI3FvrUMAL405TVxDHkAQlDYXGAAFtK4uLk8g6WTFVUjMcQN84sra7jy0mA5BcUMAIRtdGh9DAO3/i90JcO1437VWrTPIANRg12gvAAXjBCptEh8yoCavsDlEwRCoORSJ4ADwJAB8DAUAFkmgw2hiMNBsZ5hLBsBgYLIwNhJNgoABVMzYVIwbKI35oqk0nH4nhEmFQCAUhgirFiwksABaMggQoO8tp4owRIUbI4MAMqQAKlBDbKtTj6TBGczWeyuTy+dl+tsfhNkZrwdSFXjCQwYABHVXS2WyXAwUjCURiDkwACiQY4bOypKaLylEA9XpGVv9EoYXGDoZlbQjUZj4njSZTkmyJbVmbVOd2eZ9ooLuoYsFwmUS5cj0YAglw4cWQ03iQajabzTBW7mgQ0K9GRFwFDIMCaIE5/KRsr3+zmRv8Ym3osCtvLobDyPD3ReoqjhR2/Tq9WS7pS37SbXaWTZDluVwXl+UFJ8KnzD8GCzS1f0VCUVTVDV20xbUA31SRDWNM0LR/dDrSEBkmUAx0QLA103ndXxIJiaCA0bMNB0rERq1gWtU3TYpm2lE80TlBCuyJI8snDIcWFHQoADkIByJiIBeLCcLnQ1FwE+oGFXGMNy3Hc9wPUSMH4mp/kggBtEkYAwZYigMcR6ysmy7IcgB5MQ+AgMxJIAcx83tzE3IxJFMUwfJKABdSCxBMfUmW0JAZi2E4llWVQLhgK4cmACAIEkBhIDHXhPK4F48mU2c8JgMqtgq40ADE2UjJERhfA4jgYbJsjgCi+Uk3BcRyvKXmAABPJkCUPNICq8wpit4LyShKBgADJOrq1SYGojbGskZrloAamnbDZ12yNUOif5RiAA=)